### PR TITLE
[Memory Engine] add core column related classes

### PR DESCRIPTION
**Pull request by decster**
This CL add core column related classes into the code base, including:

- Buffer
- ColumnBlock
- ColumnDelta
- Column
- DeltaIndex
- ColumnSchema and Schema

Column read/write will added in the following PR

Part of #3465

### CODE REVIEW

File: be/src/olap/memory/buffer.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+// under the License.
+
+#pragma once
+
+#include "olap/memory/common.h"
+
+namespace doris {
+namespace memory {
+
+class Buffer {
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/buffer.h:29]: fixed

File: be/src/olap/memory/CMakeLists.txt [hunk 2]
```diff
@@ -22,6 +22,12 @@ set(LIBRARY_OUTPUT_PATH "${BUILD_DIR}/src/olap/memory")
 set(EXECUTABLE_OUTPUT_PATH "${BUILD_DIR}/src/olap/memory")
   
 add_library(Memory STATIC
+	buffer.cpp
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/CMakeLists.txt]: fixed

File: be/src/olap/memory/column.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+namespace doris {
+namespace memory {
+
+Column::Column(const ColumnSchema& cs, ColumnType storage_type, uint64_t version)
+        : _cs(cs), _storage_type(storage_type), _base_idx(0) {
+    _base.reserve(64);
+    _versions.reserve(64);
+    _versions.emplace_back(version);
+    DLOG(INFO) << "create %s" << debug_string();
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/column.cpp:28]: This log is only used in debug mode, like DCHECK, it help show more info when debugging, and it has no effect in release build. I can remove them if it's this project's convention.

File: be/src/olap/memory/delta_index.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        if (bid < _block_ends.size()) {
+            *start = bid > 0 ? _block_ends[bid - 1] : 0;
+            *end = _block_ends[bid];
+        } else {
+            *start = 0;
+            *end = 0;
+        }
+    }
+
+    // Return true this index has any rowid belonging to this block
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/delta_index.h]: fixed

**Comment by kangkaisen** [comment]
@decster Hi, I have one question:  Do we have a data structure to describe `the block of rows for multi columns`?  Like  Doris `RowBlockV2`, Arrow `RecordBatch` and TiDB `Chunk`.

File: be/src/olap/memory/column.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+// under the License.
+
+#include "olap/memory/column.h"
+
+namespace doris {
+namespace memory {
+
+Column::Column(const ColumnSchema& cs, ColumnType storage_type, uint64_t version)
+        : _cs(cs), _storage_type(storage_type), _base_idx(0) {
+    _base.reserve(64);
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/column.cpp:25]: > @decster Hi, I have one question: Do we have a data structure to describe `the block of rows for multi columns`? Like Doris `RowBlockV2`, Arrow `RecordBatch` and TiDB `Chunk`.

No, but the same block id and rowid represent same group of rows, so it's easy to add them if it's really necessary. But currently, I do not see any needs.

File: be/src/olap/memory/column_block.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+public:
+    ColumnBlock() = default;
+
+    size_t memory() const;
+
+    Buffer& data() { return _data; }
+
+    Buffer& nulls() { return _nulls; }
+
+    Status alloc(size_t size, size_t esize);
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/column_block.h:37]: Sure

File: be/src/olap/memory/buffer.h [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    void set_zero();
+
+    // return true if this buffer is not empty
+    operator bool() const { return _data != nullptr; }
+
+    // returns a direct pointer to the memory array
+    const uint8_t* data() const { return _data; }
+
+    // returns a direct pointer to the memory array
+    uint8_t* data() { return _data; }
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/buffer.h:50]: There are already many places use this naming convention, and class in std also use this convention.

File: be/src/olap/memory/column.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/column.cpp:25]: fixed

File: be/src/olap/memory/column_block.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/column_block.h:37]: fixed

**Comment by decster** [comment]
> @decster Hi, I have one question: Do we have a data structure to describe the block of rows for multi columns? Like Doris RowBlockV2, Arrow RecordBatch and TiDB Chunk.

Not currently, but same blockid represents same group of rows, so we can always add them latter if needed.

File: be/src/olap/memory/column.h [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    Status write(std::unique_ptr<ColumnWriter>* writer);
+
+private:
+    ColumnSchema _cs;
+    // For some types the storage_type may be different from actual type from schema.
+    // For example, string stored in dictionary, so column_block store a integer id,
+    // and the storage type may change as the dictionary grows, e.g. from uint8 to uint16
+    ColumnType _storage_type;
+    // base's position at _versions vector
+    ssize_t _base_idx;
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/column.h:74]: This is for usage in Column.cpp, but I guess change code in for loop would be better. 

`        for (ssize_t i = static_cast<ssize_t>(_base_idx) - 1; i >= 0; i--) {
`

fixed

File: be/src/olap/memory/column.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/column.cpp]: It's the step size for base vector and version vector when over capacity and need to grow

File: be/src/olap/memory/buffer.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/olap/memory/buffer.h:50]: OK

**Review by kangkaisen — DISMISSED** [review]
+1, LGTM

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/memory/common.h [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+namespace doris {
+namespace memory {
+
+template <class T, class ST>
+inline T Padding(T v, ST pad) {
+    return (v + pad - 1) / pad * pad;
+}
+
+template <class T, class ST>
+inline size_t NBlock(T v, ST bs) {
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/common.h]: fixed

File: be/src/olap/memory/common.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+#include "gutil/stringprintf.h"
+#include "olap/olap_common.h"
+#include "olap/olap_define.h"
+#include "olap/types.h"
+
+namespace doris {
+namespace memory {
+
+template <class T, class ST>
+inline T Padding(T v, ST pad) {
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/common.h]: fixed

File: be/src/olap/memory/schema.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+    std::string type_name() const;
+    std::string debug_string() const;
+
+private:
+    TabletColumn _tcolumn;
+};
+
+// Memory engine's tablet schema, simple wrapper of TabletSchema.
+// Schema have some differences comparing to original TablteSchema:
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/memory/schema.h]: fixed

**Review by chaoyli — APPROVED** [review]

**Comment by chaoyli** [comment]
LGTM

**Pull request merged**